Functors (#27)
smlhelp/book: src/concepts/functors.md

components of some software can fit together. We have used language like
"swapping out" modules or "substituting" modules for one another, but this is
very imprecise. What exactly do we mean when we say to swap one module for
another? Certainly, it would be messy to have to go through our code and change
every mention of `StructureA` for `StructureB`, and we would like to avoid some
kind of global change that requires an extenuous amount of effort on the
client's part.

Going through and changing every mention of a particular structure in order to
achieve "modularization" is akin to saying that using a particular higher-order
function by replacing specifically what each function parameter is in each
invocation is how we can achieve "parameterization" in the function argument. In
reality, this does not offer any more versatility. In this chapter, we will
discuss _functors_, which are akin to _higher-order modules_, which are allowed
to take other modules in as arguments.

## Functors: The Basics

Firstly, consider the signature of a stack.

```sml
signature STACK =
sig
    type 'a t

    val push : 'a t -> 'a -> 'a t

>     val pop : 'a t -> 'a * 'a t

Reviewer: ```suggestion
    val pop : 'a t -> 'a option * 'a t
```
Seems like this ought to be an option instead

    val size : 'a t -> int
    val empty : 'a t
end

```

Suppose that we would like to _extend_ the definition of a stack such that it
has a limit on how many elements that it contains. However, we don't necessarily
know which implementation of a stack to use - there could be several such
existing implementations, so we would like to instead make our `BoundedStack`
structure a _parameter of_ an existing structure ascribing to `STACK`.

We first consider the case where we have a hard limit of 10 items in a given
stack.

```sml
functor BoundedStack (S : STACK) :> STACK =
struct
    type 'a t = 'a S.t

    val limit = 10

>     exception Full

Reviewer: Not a huge fan of showing an example where there's an exception that is raised but not exposed in the signature. Could we have a `BOUNDED_STACK` signature that `include`s the `STACK` signature (or just repeats it if we want to avoid `include`).

    fun push S x =
        if S.size S >= limit
          then raise Full
          else S.push S x

    fun pop S = S.pop S

    fun size S = S.size S

    fun empty S = S.empty
end
```

We see that most of this code is duplicated - we have based the majority of the
design of this `BoundedStack` in terms of `S`, which is the given implementation
of a stack. The correctness of a `BoundedStack` is thus totally dependent on
whether or not the given `S` is correct, but we achieve _modularity_ in that we
can freely swap out a structure ascribing to `STACK` for another when
instantiating a given `BoundedStack`.

To put it concretely, suppose that we have two structures:

```sml
structure Stack1 :> STACK =
struct
    (* some code here *)
end

structure Stack2 :> STACK =
struct
    (* some code here *)
end
```

Then we can create instances of the `BoundedStack` functor as follows:

```sml
structure BoundedStack1 = BoundedStack(Stack1)

> structure BoundedStack2 = BoundedStack(Stack2)

Reviewer: When talking about using functors it might be worth mentioning that SML functors are generative (versus [applicative](http://caml.inria.fr/pub/papers/xleroy-applicative_functors-popl95.pdf), as they are in OCaml). Specifically we don't need to use those terms, but just mention that applying a given functor to the same structure produces two distinct structures as a result. Not sure where the best spot to put a mention of that would be though.

```

Note that `BoundedStack1` and `BoundedStack2` are themselves structures
ascribing to `STACK`, so they could be passed in as inputs to the `BoundedStack`
functor. It is not clear why you would want to do this, however.

`BoundedStack1` and `BoundedStack2` are valid stacks in their own right, so they
all can access the fields of the `STACK` signature, same as any other structure
implementing `STACK`. Presumably, the only change they display from `Stack1` and
`Stack2` are in raising the exception `Full` when the stack is given more than
ten elements.

## Functors: Syntactic Sugar

> The New Jersey implementation of Standard ML (SML/NJ, the one you are likely
> using) offers some "syntactic sugar" with respect to functors. It is a little

Reviewer: This is not an SML/NJ extension, but rather part of the definition.
```suggestion
SML offers some "syntactic sugar" for functor arguments, allowing us to
(seemingly) parameterize them by terms other than structures. It is a little
```

Certain types have some functionality or operation in common. Depending on the
operation in question, we can say that these types fall into the same
_typeclass_, which is a common interface consisting of a type and the desired
operations. Note that typeclass membership is not a formally defined
relationship, but instead a useful categorization that we use in order to
classify types that we intend to parameterize some implementation over.

For a concrete example of a typeclass, consider the `ORDERED` typeclass.

```sml
signature ORDERED =
sig
    type t
    val compare : t * t -> order
end
```

The `ORDERED` typeclass consists of those types that admit a _sensible
ordering_, which we will not (and perhaps cannot) define. Thus, we can witness
`int` and `string` as valid instances of the `ORDERED` typeclass with the
following structures:

```sml
structure IntOrder : ORDERED =
struct
    type t = int
    val compare = Int.compare
end

structure StringOrder : ORDERED =
struct
    type t = string
    val compare = String.compare
end
```

Note that it is useful, in this case, for our instances of the `ORDERED`
typeclass to be _transparently ascribed_, since it is the whole point that we

> are aware of the type that the typeclass is associated with.

Reviewer: I would add a short blurb about `where` clauses here; we really just use transparent ascription as a way to avoid talking about them, and they probably warrant at least a sentence here.

These definitions come very naturally from the fact that the `String` and `Int`
libraries included with the Standard ML basis library already implement
`compare` fields, however we can also define types such as `int list` to be
instances of `ORDERED`:

```sml
structure IntListOrder : ORDERED =
struct
    type t = int list
    fun compare [] [] = EQUAL
      | compare [] ys = LESS
      | compare xs [] = GREATER
      | compare (x::xs) (y::ys) =
        case Int.compare (x, y) of
            EQUAL => compare xs ys
          | LESS => LESS

>           | GREATER => GREATER

Reviewer: Can be simplified to

```sml
fun compare xs ys = List.collate Int.compare (xs, ys)
```

though if we want to be explicit this is fine. Same goes for the generic version.

Reviewer: Also this shouldn't be curried, so it can even just be `val compare = List.collate Int.compare`

Reviewer 2: Do we commonly teach `List.collate`? If not, it might be worth explaining its function elsewhere (and perhaps using exactly this `compare` function as an example) and linking to it.

signature:

```sml
signature EQ =
sig
    type t
    val equal : t * t -> bool
end
```

The operations for `equal` for each "instance" of the typeclass are instead
defined by Standard ML itself, and not user-defined. Thus, we can think of the
equality operator `=` as simply invoking the `T.equal` method for the proper
typeclass `T`, defined by the type that is being compared for equality.

In this next section, we will explore a concrete use for typeclasses when
designing functors.

## Case study: Red-black trees

Typeclasses can be important when we are attempting to place some greater
constraint on the types that may instantiate some universal type. In certain
cases, we do not want the types that we are considering to truly be _any_ type,
but any of a limited subset of types that share some common characteristic or
implement some operation. We will study the use of _red-black trees_ as the
underlying data structure for dictionaries.

A dictionary is a simple data structure that maps keys to values. Consider its
signature given below.

```sml
signature DICT =
sig
    type key
    type 'a dict
    val empty : 'a dict
    val insert : 'a dict -> key * 'a -> 'a dict

>     val find : 'a dict -> key -> 'a option

Reviewer: Rename for consistency---`find` and `lookup` are both used. I prefer `lookup` so suggestions standarise on that.

```suggestion
    val lookup : 'a dict -> key -> 'a option
```

end
```

It is a well-known fact that, utilizing a kind of _balanced binary tree_ data
structure, dictionaries can be implemented with an \\( O(\log n) \\) `insert` and

> `find` operation, as opposed to \\( O(n) \\) for other data structures such as

Reviewer: Ibid.
```suggestion
`lookup` operation, as opposed to \\( O(n) \\) for other data structures such as
```

    case Key.compare (k, k') of
        LESS => balance(Node(insert L (k, v), (c', (k', v')), R))
      | EQUAL => Node(L, (c', (k, v)), R)
      | GREATER => balance(Node(L, (c', (k', v')), insert R (k ,v)))
```

This code ensures that, after descending into a subtree in order to insert the
given key and value, a balancing operation is immediately performed once the
insertion is complete. This ensures that we have a _bottom-up_ propagation of
balancings, immediately after completing the insertions. Note that because
`balance` acts as the identity function on anything that does not pattern-match
to either of the four cases, we perform only a negligible amount of extra checks
at each recursive call, and ultimately are only concerned with those four cases.

However, this code is not complete. There is a minor edge case that remains -
what if we are too close to the root for any of the four cases to apply? Our
previous analysis relied on the fact that we could assume that the parent of our
inserted node was red, and thus had a black parent - what of the case where the
parent of the inserted node _has no_ parent?

Consider the case illustrated below:

> some illustration of the red-red insertion case goes here

Reviewer: ```suggestion
<figure class="aligncenter">
    <img src="../assets/insert.svg" alt="Example of inserting two nodes into an empty tree" width="1000"\>
    <figcaption><b>Fig 3.</b> Demonstration of potential issues in inserting nodes at the root when lacking a black parent. </figcaption>
</figure>
```

however, since that is the only location where that can occur. As a result, the
simple solution is to simply make the root of any red-black tree black - it will
preserve the black height invariant, but also result in this red-red violation
being impossible. We can amend our code as follows:

```sml
fun insert' Leaf (k, v) = Node(Leaf, (Red, (k, v)), Leaf)
  | insert' (Node (L, (c', (k', v')), R)) (k, v) =
    case Key.compare (k, k') of
        LESS => balance(Node(insert' L (k, v), (c', (k', v')), R))
      | EQUAL => Node(L, (c', (k, v)), R)
      | GREATER => balance(Node(L, (c', (k', v')), insert' R (k ,v)))

fun insert T (k, v) =
    case insert' T (k, v) of
        Leaf => Leaf
      | Node (L, (_, (k', v')), R) => Node(L, (Black, (k', v')), R)
```

Finally, this results in our completed code for the `insert` function. Note that
because of the signature that we are ascribing to, helper functions such as
`balance` and `insert` will not be visible to the client of the module, so there
is no harm in declaring them within the namespace of the functor.

Our completed code for a red-black tree implementation of dictionaries is thus

> as follows. Note that the implementation of `find` is very straightforward, and

Reviewer: Ibid.
```suggestion
as follows. Note that the implementation of `lookup` is very straightforward, and
```
